opal/util/opal_gethostname.c

```diff
@@ -0,0 +1,173 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * 
+ * Copyright (c) 2019      Triad National Security, LLC.
+ *                         All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ * Based on pgethostname (a portable gethostname)
+ * available at http://perfec.to/gethostname/
+ * and Graham Shaw's sample code at
+ * http://www.microhowto.info/howto/determine_the_fully_qualified_hostname_of_the_local_machine_in_c.html
+ *
+ * $HEADER$
+ */
+
+#ifndef TESTING
+#include "opal/util/opal_gethostname.h"
+#else
+#include <stdbool.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#if defined(HAVE_UNISTD_H) || defined(TESTING)
+#include <unistd.h>
+#endif
+
+/*
+ * If there is a preprocessor macro that redefined the call to
+ * gethostname, we undefine that here 
+ */
+#ifdef gethostname
+#undef gethostname
+#endif
+
+/* Build with -DTESTING=1 if we are building small testing standalone (with test_opalgethostname.c) */
+#ifdef TESTING
+#define gethostname(a, b) dummy_gethostname((a), (b))
+#define OPAL_LOCAL_MAXHOSTNAMELEN 4096
+bool first_pass = true;
+
+/* dummy_gethostname function used for testing */
+int dummy_gethostname( char *, size_t );
+
+#else /* this is for a production build, not a standalone test build, so we use the real gethostname() and declare first_pass as static */
+static bool first_pass = true;
+#endif /* ifdef TESTING */
+
+
+#define NUM_TRIES_FOR_NULL_HOSTNAME 8
+
+#define MEMORY_ERROR_STRING "Invalid address (EFAULT) in call to gethostname. Check memory usage."
+#define HOSTNAMELEN_ERROR_STRING "Error in call to gethostname() - check hostname length. Excessively long hostnames are unsupported."
+#define UNKNOWN_ERROR_STRING "Unknown error in opal_gethostname() function. Probable code issue."
+/*
+ * This gethostname wrapper does not return the full-length hostname in
+ * those cases where it is too long for the buffer. It does, however,
+ * guarantee a null-terminated hostname is returned, even if it's
+ * truncated. It also tries again in the case where gethostname returns an 
+ * error because the buffer is too short, but once it has retrieved the full
+ * name without error, it truncates the result and resets errno to 
+ * ENAMETOOLONG. 
+ */
+static char opal_local_hostname[OPAL_LOCAL_MAXHOSTNAMELEN] =
+    HOSTNAMELEN_ERROR_STRING;
+
+
+const char *opal_gethostname( void )
+{
+    size_t count;
+    size_t length = OPAL_LOCAL_MAXHOSTNAMELEN;
+    int ret_val, num_tries = 0;
+    int temp_errno = 0;
+
+    if( first_pass ) {
```

Comment:
Are we worried about thread safety here? It's conceivable that two threads can be in here at the same time for first the first_pass case. Might lead to some goofy behavior.

Reply:
I'm going to let others comment on whether or not this needs to be reentrant. I am new to this effort and did not code it with that in mind. Having said that, I followed @jsquyres 's suggestion on this in the original issue - open-mpi#6801. Perhaps you should read that issue to catch up. A comment there from you earlier might have been helpful.

Reply:
I'm not so worried about reentrantcy as we are of two threads calling this at the same time, and each thread executing the `first_pass` block at the same time. Considering the scope of this change, I'd bet that it is *possible*, but maybe not probable.

Reply:
I think this concern could be addressed by a statically initialized mutex to protect accessing ```first pass```

Reply:
Ok, if statically initialized mutexes are already used elsewhere in other parts of the code, I'll make the necessary change.

Reply:
Should be resolved.